### PR TITLE
Extract PCG RandomNumberGenerator

### LINKED ISSUE #37: Polish fast RNG & make available publicly
https://github.com/saeta/penguin/pull/30 pulled the `PCGRandomNumberGenerator` out from `NonBlockingThreadPool.swift`, but it (and `fastFit`) should really move to `PenguinStructures` (or some similar library) and be made publicly available (and generic where appropriate).

Note: https://github.com/saeta/penguin/pull/36 pulls out the other numerical bits and cleans them up.

### PR DESCRIPTION
**Pull request by robb**
Here's my first pass at #37.

The original implementation of `PCGRandomNumberGenerator` didn't adhere to `RandomNumberGenerator` correctly since it didn't actually implement the `UInt64` variant of `next`, see also the test in 1995d34.

Some notes:
- I modeled my implementation after [an example in the reference implementation](https://github.com/imneme/pcg-c/blob/83252d9c23df9c82ecb42210afed61a7b42402d7/sample/pcg32x2-demo.c) with the only difference being that this still uses the PCG-XSH-RS variant, not PCG-XSH-RR. Alternatively, we could also implement a 128 bit state, 64 bit output variant but this seems simpler.
- I've struggled a bit with naming on the `PCG_XSH_RS_32_Generator` struct, I found `PCGXSHRS32Generator` hard to parse and as established, it doesn't conform to `RandomNumberGenerator` by itself. What do you think?
- I noticed that there is no precedence for `@testable` imports in the PenguinStructuresTests, I'm happy to make the underlying RNG public and keep it that way, too.

(Also, :wave: @dabrahams – long time no see)

Fixes #37.

### CODE REVIEW

File: Sources/PenguinStructures/PCGRandomNumberGenerator.swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  public init(state: UInt64) {
+    self.init(seed: state, seq: 0xda3e_39cb_94b9_5bdb)
+  }
+
+  public init(seed: UInt64, seq: UInt64) {
+    self.init(seed1: seed, seed2: seed, seq1: seq, seq2: seq)
+  }
+
+  public init(seed1: UInt64, seed2: UInt64, seq1: UInt64, seq2: UInt64) {
```
Comment (robb) [inline Sources/PenguinStructures/PCGRandomNumberGenerator.swift:34]: Not super happy with these parameter names.

**Comment by saeta** [comment]
Thank you @robb for putting this together! I'll take a look shortly.

File: Sources/PenguinParallel/NonblockingThreadPool/NonBlockingThreadPool.swift [hunk 2]
```diff
@@ -467,9 +467,9 @@ fileprivate final class PerThreadState<Environment: ConcurrencyPlatform> {
   var isCancelled: Bool { pool.cancelled }
 
   func steal() -> Task? {
-    let r = Int(rng.next())
-    var selectedThreadId = fastFit(r, into: pool.totalThreadCount)
-    let step = pool.coprimes[fastFit(r, into: pool.coprimes.count)]
+    let r = rng.next() as UInt32
```
Comment (saeta) [inline Sources/PenguinParallel/NonblockingThreadPool/NonBlockingThreadPool.swift:470]: Do you still need the `as UInt32`?

File: Sources/PenguinStructures/PCGRandomNumberGenerator.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    let mask: UInt64 = ~0 >> 1;
+
+    var (stream1, stream2) = (seq1, seq2)
+
+    if stream1 & mask == stream2 & mask {
```
Comment (saeta) [inline Sources/PenguinStructures/PCGRandomNumberGenerator.swift:39]: Sorry, I'm not familiar with this construction. Although I understand what the code-as-written is doing, do you mind adding a comment explaining the larger context for this approach, what properties this provides, and what guarantees need to be upheld?

File: Sources/PenguinStructures/PCGRandomNumberGenerator.swift [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      stream2 = ~stream2
+    }
+
+    first  = PCG_XSH_RS_32_Generator(state: seed1, stream: stream1)
+    second = PCG_XSH_RS_32_Generator(state: seed2, stream: stream2)
+  }
+
+  public mutating func next() -> UInt64 {
+    UInt64(truncatingIfNeeded: first.next()) << 32 | UInt64(truncatingIfNeeded: second.next())
```
[the remark this answers is not in the document]
Comment (robb) [reply Sources/PenguinStructures/PCGRandomNumberGenerator.swift]: You're right. I somehow thought that this would be slightly faster but a quick benchmark doesn't back that up.

File: Sources/PenguinParallel/NonblockingThreadPool/NonBlockingThreadPool.swift (hunk 2 above)
Comment (robb) [reply Sources/PenguinParallel/NonblockingThreadPool/NonBlockingThreadPool.swift:470]: Nope, one more fix and then this should be ready to go.

**Comment by robb** [comment]
🌴
